File: typescript/src/model/expressions/wdl-string-literal.ts
/** String literal and placeholder expression nodes. */
import { WdlExpressionComponentType, type WdlExpression } from './wdl-expression.js';

/** Source-level string delimiter forms supported by WDL. */
export enum Delimiter {
  SINGLE_QUOTED = '"',
  DOUBLE_ANGLE = '<<<>>>',
}

/** Component kinds that may appear inside a string literal. */
export enum WdlStringComponentType {
  TEXT = 'TEXT',
  ESCAPE = 'ESCAPE',
  TOKEN = 'TOKEN',
  PLACEHOLDER = 'PLACEHOLDER',
}

/** Base interface implemented by fragments inside a parsed string literal. */
export interface WdlStringComponent {
  /** Returns the broad string-component family for this node. */
  componentType(): WdlStringComponentType;
}

/** Plain text fragment inside a string literal. */
export class WdlStringText implements WdlStringComponent {
  /** Creates a text fragment from its source text. */
  public constructor(public text?: string) {}
  /** Returns the string-component family for this node. */
  public componentType(): WdlStringComponentType {
    return WdlStringComponentType.TEXT;
  }
}

/** Escape fragment inside a string literal. */
export class WdlStringEscape implements WdlStringComponent {
  /** Creates an escape fragment from its escaped source text. */
  public constructor(public escapeText?: string) {}
  /** Returns the string-component family for this node. */
  public componentType(): WdlStringComponentType {
    return WdlStringComponentType.ESCAPE;
  }
}

/** Token-preserving fragment inside a string literal. */
export class WdlStringToken implements WdlStringComponent {
  /** Creates a token fragment from its token text. */
  public constructor(public tokenText?: string) {}
  /** Returns the string-component family for this node. */
  public componentType(): WdlStringComponentType {
    return WdlStringComponentType.TOKEN;
  }
}

/** Supported placeholder sigils used in string interpolation. */
export enum PlaceHolderSymbol {
  TILDE = '~',
  DOLLAR = '$',
}

/** Supported placeholder option families from the WDL grammar. */
export enum WdlStringPlaceholderOptionType {
  DEFAULT = 'default',
  TRUE_FALSE = 'true_false',
}

/** Option block attached to a string placeholder. */
export class WdlStringPlaceholderOption {
  /** Creates placeholder options from the option type and optional string-literal payloads. */
  public constructor(
    public type: WdlStringPlaceholderOptionType,
    public value?: WdlStringLiteral,
    public trueValue?: WdlStringLiteral,
    public falseValue?: WdlStringLiteral,
  ) {}
}

/** Interpolated placeholder fragment inside a string literal. */
export class WdlStringPlaceholder implements WdlStringComponent {
  /** Creates a placeholder from its options, expression, and sigil. */
  public constructor(
    public option?: WdlStringPlaceholderOption,
    public expression?: WdlExpression,
    public symbol: PlaceHolderSymbol = PlaceHolderSymbol.TILDE,
  ) {}

  /** Returns the string-component family for this node. */
  public componentType(): WdlStringComponentType {
    return WdlStringComponentType.PLACEHOLDER;
  }
}

/** Models a WDL string literal assembled from ordered fragments. */
export class WdlStringLiteral implements WdlExpression {
  private readonly componentValues: WdlStringComponent[] = [];

  /** Creates a string literal from its delimiter form. */
  public constructor(public delimiter: Delimiter = Delimiter.SINGLE_QUOTED) {}

  /** Returns the ordered fragments that make up the string literal. */
  public components(): WdlStringComponent[] {
    return this.componentValues;
  }

  /** Returns the broad expression family for this node. */
  public componentType(): WdlExpressionComponentType {
    return WdlExpressionComponentType.STR_LIT;
  }
}

File: python/src/wdl_model/model/expressions/wdl_string_literal.py
"""String literal and placeholder expression nodes.

WDL strings are richer than plain scalar literals because they can contain interpolation
placeholders and command-template fragments.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from enum import Enum

from .wdl_expression import ComponentType, WdlExpression


class Delimiter(Enum):
    """Source-level string delimiter forms used by WDL literals and command templates."""

    SINGLE_QUOTED = '"'
    DOUBLE_ANGLE = "<<<>>>"


class WdlStringComponent:
    """Base interface for a component inside a parsed WDL string literal."""

    class ComponentType(Enum):
        TEXT = "TEXT"
        ESCAPE = "ESCAPE"
        TOKEN = "TOKEN"
        PLACEHOLDER = "PLACEHOLDER"

    def componentType(self) -> "WdlStringComponent.ComponentType":
        raise NotImplementedError


@dataclass
class WdlStringText(WdlStringComponent):
    """Raw text fragment inside a string literal."""

    text: str | None = None

    def componentType(self) -> WdlStringComponent.ComponentType:
        return WdlStringComponent.ComponentType.TEXT


@dataclass
class WdlStringEscape(WdlStringComponent):
    """Escape sequence fragment inside a string literal."""

    escapeText: str | None = None

    def componentType(self) -> WdlStringComponent.ComponentType:
        return WdlStringComponent.ComponentType.ESCAPE


@dataclass
class WdlStringToken(WdlStringComponent):
    """Token-preserving fragment used when rebuilding string text."""

    tokenText: str | None = None

    def componentType(self) -> WdlStringComponent.ComponentType:
        return WdlStringComponent.ComponentType.TOKEN


class PlaceHolderSymbol(Enum):
    """Supported placeholder sigils for WDL string interpolation."""

    TILDE = "~"
    DOLLAR = "$"

    def getWdlString(self) -> str:
        return self.value


class Type(Enum):
    """Placeholder option families supported by older WDL placeholder syntax."""

    DEFAULT = "default"
    TRUE_FALSE = "true_false"


@dataclass
class WdlStringPlaceholderOption:
    """Option block attached to a string placeholder, such as default or true/false."""

    type: Type
    value: "WdlStringLiteral" | None = None
    trueValue: "WdlStringLiteral" | None = None
    falseValue: "WdlStringLiteral" | None = None

    def getType(self) -> Type:
        return self.type

    def getValue(self) -> "WdlStringLiteral" | None:
        return self.value

    def setValue(self, value: "WdlStringLiteral") -> None:
        self.value = value

    def getTrueValue(self) -> "WdlStringLiteral" | None:
        return self.trueValue

    def setTrueValue(self, value: "WdlStringLiteral") -> None:
        self.trueValue = value

    def getFalseValue(self) -> "WdlStringLiteral" | None:
        return self.falseValue

    def setFalseValue(self, value: "WdlStringLiteral") -> None:
        self.falseValue = value


@dataclass
class WdlStringPlaceholder(WdlStringComponent):
    """Interpolated placeholder embedded in a string literal or command template."""

    option: WdlStringPlaceholderOption | None = None
    expression: WdlExpression | None = None
    symbol: PlaceHolderSymbol = PlaceHolderSymbol.TILDE

    def componentType(self) -> WdlStringComponent.ComponentType:
        return WdlStringComponent.ComponentType.PLACEHOLDER


@dataclass
class WdlStringLiteral(WdlExpression):
    """String literal expression composed of ordered text and placeholder fragments."""

    delimiter: Delimiter = Delimiter.SINGLE_QUOTED
    _components: deque[WdlStringComponent] = field(default_factory=deque)

    def components(self) -> deque[WdlStringComponent]:
        """Return the ordered string components that make up the literal."""
        return self._components

    def componentType(self) -> ComponentType:
        return ComponentType.LITERAL

    def __str__(self) -> str:
        chunks: list[str] = []
        for component in self._components:
            if isinstance(component, WdlStringText):
                chunks.append(component.text or "")
            elif isinstance(component, WdlStringEscape):
                chunks.append(component.escapeText or "")
            elif isinstance(component, WdlStringToken):
                chunks.append(component.tokenText or "")
            elif isinstance(component, WdlStringPlaceholder):
                chunks.append(
                    f"{component.symbol.value}{{{component.expression or ''}}}"
                )
        return "".join(chunks)
